atomic_reactor/plugins/pre_reactor_config.py

```diff
@@ -0,0 +1,363 @@
+                return errors
+
+            to_check = conf.keys()
+        elif isinstance(element, list):
+            # Check every value at this level
+            to_check = range(len(conf))
+        else:
+            to_check = [element]
+
+        for each_key in to_check:
+            subpath = path + (each_key,)
+            try:
+                subconf = conf[each_key]
+            except KeyError:
+                if len(required) > 1:
+                    # Not final element, ignore
+                    continue
+
+                errpath = '.'.join(subpath)
+                errors.append("{} missing".format(errpath))
+            except TypeError:
+                errpath = '.'.join(path)
+                errors.append("{} must be map".format(errpath))
+            else:
+                errs = self.validate_required_path(subconf,
+                                                   required=required[1:],
+                                                   path=subpath)
+                errors.extend(errs)
+
+        return errors
+
+
+class ClusterConfig(object):
+    """
+    Configuration relating to a particular cluster
+    """
+
+    def __init__(self, name, max_concurrent_builds=None, enabled=True):
+        self.name = str(name)
+        self.max_concurrent_builds = None
+        self.enabled = enabled
+
+        if max_concurrent_builds is not None:
+            self.max_concurrent_builds = int(max_concurrent_builds)
+
+
+class ReactorConfig(object):
+    """
+    Class to parse the atomic-reactor configuration file
+    """
+
+    CLUSTERS_KEY='clusters'
+    SCHEMA = {
+        'clusters': {
+            str: [
+                {
+                    'name': str,
+                    'max_concurrent_builds': int,
+                    'enabled': bool,
+                },
+            ]
+        }
+    }
+
+    REQUIRED = [
+        # clusters.*.name is required (but 'clusters' is not)
+        ('clusters', None, ['name']),
+    ]
+
+    def __init__(self, config_path=None, log=None):
+        self.log = log or logging.getLogger(__name__)
+
+        if config_path is None:
+            self.conf = {}
+            return
+
+        self.log.info("reading config from %s", config_path)
+        with open(config_path) as fp:
+            conf = yaml.safe_load(fp)
+
+        config_builder = ConfigBuilder(schema=self.SCHEMA,
+                                       required=self.REQUIRED)
+        errors, conf = config_builder.build_conf(conf)
+        errors.extend(config_builder.validate_required(conf))
+        if errors:
+            for error in errors:
+                self.log.error(error)
+
+            raise RuntimeError('config validation failed')
+
+        self.conf = conf
+
+    def get_enabled_clusters_for_platform(self, platform):
+        clusters = self.conf.get(self.CLUSTERS_KEY, {}).get(platform, [])
+        cluster_configs = [ClusterConfig(**cluster) for cluster in clusters]
```

Comment:
Would it be better to only create these `ClusterConfig` objects once? Seems that the use case is for  `get_enabled_clusters_for_platform` to be called for each platform.